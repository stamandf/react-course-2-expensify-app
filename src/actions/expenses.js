import uuid from 'uuid';

// ADD_EXPENSE Action
export const addExpense = ({description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

export const removeExpense = ({id} = {}) => ({
    type: 'DELETE_EXPENSE',
    id
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});
