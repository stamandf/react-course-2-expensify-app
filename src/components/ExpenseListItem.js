import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    
        <li><Link to={"/edit/"+id}>{description}</Link>: ${amount} date:{createdAt}</li>
    )

export default ExpenseListItem;