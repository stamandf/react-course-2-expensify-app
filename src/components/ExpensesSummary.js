import React from 'react';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';
import { connect } from 'react-redux';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    if (expensesCount === 0) {
        return <p>No expenses.</p>
    } else {
        const expenseWord = expensesCount > 1 ? ("expenses") : ("expense");
        const totalFormatted = numeral(expensesTotal/100).format('$0,0.00'); 
        return (
            <p>
            Viewing {expensesCount} {expenseWord} totalling : {totalFormatted}.
        </p>
        )
        
    }
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length, 
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};
export default connect(mapStateToProps)(ExpensesSummary);

//Previous version:
//================================================

// export const ExpensesSummary = (props) => {
//     console.log("props:", props);
    
//     if (props.expenses.length === 0) {
//         return <p>No expenses.</p>
//     } else {
//         const expenseWord = props.expenses.length > 1 ? ("expenses") : ("expense");
//         const amount = selectExpensesTotal(props.expenses); 
//         return (
//             <p>
//             Viewing {props.expenses.length} {expenseWord} totalling : {numeral(amount/100).format('$0,0.00')}
//         </p>
//         )
        
//     }
// }

// const mapStateToProps = (state) => {
//     return {
//         expenses:  selectExpenses(state.expenses, state.filters), 
//         filters: state.filters
//     };
// };
