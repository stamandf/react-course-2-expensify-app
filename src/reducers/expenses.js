// Expenses reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {  //Action creator
    switch (action.type) {
        case 'ADD_EXPENSE':
            // return state.concat(action.expense);
            return [...state, action.expense]; //The spreading operator for arrays
        case 'DELETE_EXPENSE':
            return state.filter(({id}) => id != action.id);  
        case 'EDIT_EXPENSE':   
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        default:
            return state;
    }
};
