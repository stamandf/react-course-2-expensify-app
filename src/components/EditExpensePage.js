import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onClick = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
            <h1>Edit Expense</h1>
                 <ExpenseForm
                    expense={this.props.expense} 
                    onSubmit = {this.onSubmit}
                 />
                <button onClick={this.onClick}>Remove</button>

            </div>
        );
    };
}

// const startEditExpensePage = (props) => {
//     // console.log('Edit Expense props:',props.expense);
//     return (
//         <div>
//         <h1>Edit Expense</h1>
//              <ExpenseForm 
//                 expense={props.expense}
//                  onSubmit={(expense) => {
//                     props.dispatch(startEditExpense(props.expense.id, expense));
//                     props.history.push('/'); 
//                     // console.log('Expense updated:', expense)
//                  }}
//              />
//             <button onClick={() => {
//             props.dispatch(removeExpense({ id: props.expense.id }))  // store.dispatch(removeExpense({id: expenseOne.expense.id}));
//             props.history.push('/');
//             // console.log('Expense deleted:', props.expense.id) 
//         }} >Remove</button>
//         </div>
//     );
// };

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
        };
     };

const mapDispatchToProps = (dispatch) => ({
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    });

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
