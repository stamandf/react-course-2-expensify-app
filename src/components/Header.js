import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <div>
            <ol>
                <li><NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink></li>
                <li><NavLink to="/create" activeClassName="is-active">Create Expense</NavLink></li>
                <button onClick={startLogout}>Logout</button>
            </ol>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);