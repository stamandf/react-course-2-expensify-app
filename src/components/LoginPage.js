import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div>
        <h1>Expensify App</h1>
        <h2>Sign in</h2>
        
        <button onClick={startLogin}>Sign In</button>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);