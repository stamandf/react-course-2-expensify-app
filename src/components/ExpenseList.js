import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            <ul>
                {props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id}{...expense} />
                })}
            </ul>
            
             
        }
    
    </div>
);
// Version 1:
// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses
//     }
// })(ExpenseList);
// export default ConnectedExpenseList;

// Version 2:
// export default = connect((state) => {
//     return {
//         expenses: state.expenses
//     }
// })(ExpenseList);

//Version 3 -- The Real World way, i.e. real world code bases:
// const mapStateToProps = (state) => {
//     return {
//         expenses: state.expenses,
//         filters: state.filters
//     };
// };
const mapStateToProps = (state) => {
    return {
        expenses:  selectExpenses(state.expenses, state.filters), 
        filters: state.filters
    };
};
export default connect(mapStateToProps)(ExpenseList);

