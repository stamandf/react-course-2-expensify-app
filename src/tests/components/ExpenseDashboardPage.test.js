import React from 'react';
import { shallow } from 'enzyme';
import ExpenseDashboardPage from '../../components/ExpenseDashboardPage';
import { ExpenseList } from '../../components/ExpenseList';
import { ExpenseListFilters } from '../../components/ExpenseListFilters';
import expenses from '../fixtures/expenses';

test('should render the expense dashboard page correctly with no filter', () => {
    const wrapper = shallow(<ExpenseDashboardPage />)
    expect(wrapper).toMatchSnapshot();
});