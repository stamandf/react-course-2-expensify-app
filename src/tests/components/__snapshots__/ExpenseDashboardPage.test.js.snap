// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the expense dashboard page correctly with no filter 1`] = `
<div>
  This is from my dashboard component.
  <Connect(ExpenseListFilters) />
  <Connect(ExpensesSummary) />
  <Connect(ExpenseList) />
</div>
`;
